File: baba-chatbot-python/src/baba_chatbot/integrations/catalog/catalog_repository.py
"""
Repositorio para cargar y consultar el catálogo de vehículos desde archivos CSV y JSON.
Implementa caché en memoria para optimizar el rendimiento.
"""
import os
import json
import csv
import logging
from typing import List, Optional, Dict
from ...application.abstractions.interfaces import ICatalogRepository
from ...domain.entities import Vehicle
from ...domain.value_objects import VehicleQuery


logger = logging.getLogger(__name__)


class CatalogRepository(ICatalogRepository):
    """
    Repositorio para cargar y consultar el catálogo de vehículos.
    """

    def __init__(
        self,
        catalog_file_path: str = "./data/catalog/cars_extract.json",
        csv_file_path: str = None
    ):
        """
        Inicializa el repositorio de catálogo.
        
        Args:
            catalog_file_path: Ruta al archivo JSON del catálogo
            csv_file_path: Ruta al archivo CSV del catálogo (opcional)
        """
        self._catalog_file_path = catalog_file_path
        self._csv_file_path = csv_file_path
        self._cached_vehicles: Optional[List[Vehicle]] = None

    async def search_vehicles(self, query: VehicleQuery) -> List[Vehicle]:
        """
        Busca vehículos en el catálogo aplicando los filtros especificados en el query.
        Si no hay filtros, devuelve los primeros 10 vehículos.
        
        Args:
            query: Criterios de búsqueda
            
        Returns:
            Lista de vehículos que coinciden con los criterios
        """
        all_vehicles = await self._load_vehicles()

        if not query.has_any_filter():
            return all_vehicles[:10]

        filtered = [v for v in all_vehicles if self._matches_query(v, query)]
        return filtered

    async def get_vehicle_by_id(self, vehicle_id: str) -> Optional[Vehicle]:
        """
        Obtiene un vehículo específico por su ID o stock_id.
        
        Args:
            vehicle_id: ID o stock_id del vehículo
            
        Returns:
            Vehículo encontrado o None si no existe
        """
        vehicles = await self._load_vehicles()
        return next(
            (v for v in vehicles if v.id == vehicle_id or v.stock_id == vehicle_id),
            None
        )

    async def _load_vehicles(self) -> List[Vehicle]:
        """
        Carga los vehículos desde archivos CSV y JSON con caché en memoria.
        
        Returns:
            Lista de vehículos cargados
        """
        if self._cached_vehicles is not None:
            return self._cached_vehicles

        try:
            vehicles = []

            # Cargar desde CSV si existe
            if self._csv_file_path and os.path.exists(self._csv_file_path):
                logger.info(f"Loading vehicles from CSV: {self._csv_file_path}")
                vehicles.extend(await self._load_from_csv(self._csv_file_path))

            # Cargar desde JSON si existe
            if os.path.exists(self._catalog_file_path):
                logger.info(f"Loading vehicles from JSON: {self._catalog_file_path}")
                vehicles.extend(await self._load_from_json(self._catalog_file_path))

            if not vehicles:
                logger.warning("No catalog files found or loaded")

            self._cached_vehicles = vehicles
            logger.info(f"Loaded {len(self._cached_vehicles)} vehicles from catalog")

            return self._cached_vehicles
        except Exception as ex:
            logger.error(f"Error loading catalog: {ex}")
            return []

    async def _load_from_csv(self, file_path: str) -> List[Vehicle]:
        """
        Carga vehículos desde un archivo CSV.
        Formato esperado: stock_id,km,price,make,model,year,version,bluetooth,largo,ancho,altura,car_play
        
        Args:
            file_path: Ruta al archivo CSV
            
        Returns:
            Lista de vehículos cargados
        """
        vehicles = []
        
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                reader = csv.DictReader(f)
                
                for line_number, row in enumerate(reader, start=2):
                    try:
                        stock_id = row.get('stock_id', '').strip()
                        km = self._parse_int(row.get('km', ''))
                        price = self._parse_float(row.get('price', ''))
                        make = row.get('make', '').strip()
                        model = row.get('model', '').strip()
                        year = self._parse_int(row.get('year', '')) or 2024
                        version = row.get('version', '').strip()
                        bluetooth = self._parse_bool(row.get('bluetooth', ''))
                        length = self._parse_int(row.get('largo', ''))
                        width = self._parse_int(row.get('ancho', ''))
                        height = self._parse_int(row.get('altura', ''))
                        carplay = self._parse_bool(row.get('car_play', ''))

                        vehicle = Vehicle.create(
                            vehicle_id=stock_id,
                            brand=make,
                            model=model,
                            year=year,
                            price=price or 0.0
                        )

                        vehicle.update_details(
                            version=version,
                            mileage=km,
                            length=length,
                            width=width,
                            height=height,
                            has_bluetooth=bluetooth,
                            has_carplay=carplay,
                            stock_id=stock_id
                        )

                        vehicles.append(vehicle)
                    except Exception as ex:
                        logger.warning(f"Error parsing line {line_number}: {ex}")
        except Exception as ex:
            logger.error(f"Error reading CSV file {file_path}: {ex}")

        return vehicles

    async def _load_from_json(self, file_path: str) -> List[Vehicle]:
        """
        Carga vehículos desde un archivo JSON.
        
        Args:
            file_path: Ruta al archivo JSON
            
        Returns:
            Lista de vehículos cargados
        """
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                
            vehicles = []
            for item in data:
                vehicle = self._map_to_entity(item)
                vehicles.append(vehicle)
            
            return vehicles
        except Exception as ex:
            logger.error(f"Error reading JSON file {file_path}: {ex}")
            return []

    def _parse_int(self, value: str) -> Optional[int]:
        """
        Parsea un valor string a entero manejando valores nulos o "NA".
        
        Args:
            value: Valor a parsear
            
        Returns:
            Entero parseado o None
        """
        if not value or value.strip().upper() == 'NA':
            return None
        
        try:
            return int(float(value.strip()))
        except (ValueError, AttributeError):
            return None

    def _parse_float(self, value: str) -> Optional[float]:
        """
        Parsea un valor string a float manejando valores nulos o "NA".
        
        Args:
            value: Valor a parsear
            
        Returns:
            Float parseado o None
        """
        if not value or value.strip().upper() == 'NA':
            return None
        
        try:
            return float(value.strip())
        except (ValueError, AttributeError):
            return None

    def _parse_bool(self, value: str) -> Optional[bool]:
        """
        Parsea un valor string a booleano manejando valores nulos, "NA" o enteros (1/0).
        
        Args:
            value: Valor a parsear
            
        Returns:
            Booleano parseado o None
        """
        if not value or value.strip().upper() == 'NA':
            return None
        
        value = value.strip().lower()
        
        if value in ('true', 'yes', '1'):
            return True
        elif value in ('false', 'no', '0'):
            return False
        
        try:
            return int(value) != 0
        except (ValueError, AttributeError):
            return None

    def _matches_query(self, vehicle: Vehicle, query: VehicleQuery) -> bool:
        """
        Verifica si un vehículo cumple con los criterios del query especificado.
        
        Args:
            vehicle: Vehículo a verificar
            query: Criterios de búsqueda
            
        Returns:
            True si el vehículo cumple con los criterios
        """
        if query.brand and query.brand.lower() not in vehicle.brand.lower():
            return False

        if query.model and query.model.lower() not in vehicle.model.lower():
            return False

        if query.max_price is not None and vehicle.price > query.max_price:
            return False

        if query.min_price is not None and vehicle.price < query.min_price:
            return False

        if query.min_year is not None and vehicle.year < query.min_year:
            return False

        if query.max_year is not None and vehicle.year > query.max_year:
            return False

        if query.max_mileage is not None and vehicle.mileage > query.max_mileage:
            return False

        return True

    def _map_to_entity(self, dto: Dict) -> Vehicle:
        """
        Mapea un DTO de JSON a la entidad de dominio Vehicle.
        
        Args:
            dto: Diccionario con datos del vehículo
            
        Returns:
            Instancia de Vehicle
        """
        vehicle = Vehicle.create(
            vehicle_id=dto.get('id', ''),
            brand=dto.get('brand', ''),
            model=dto.get('model', ''),
            year=dto.get('year', 2024),
            price=dto.get('price', 0.0)
        )
        return vehicle

File: baba-chatbot-net/src/Baba.Chatbot.Integrations/Catalog/CatalogRepository.cs
using Baba.Chatbot.Application.Abstractions;
using Baba.Chatbot.Domain.Entities;
using Baba.Chatbot.Domain.ValueObjects;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using System.Globalization;
using System.Text.Json;

namespace Baba.Chatbot.Integrations.Catalog;

/// <summary>
/// Repositorio para cargar y consultar el catálogo de vehículos desde archivos CSV y JSON
/// Implementa caché en memoria para optimizar el rendimiento
/// </summary>
public class CatalogRepository : ICatalogRepository
{
    private readonly ILogger<CatalogRepository> _logger;
    private readonly string _catalogFilePath;
    private readonly string _csvFilePath;
    private List<Vehicle>? _cachedVehicles;

    /// <summary>
    /// Inicializa una nueva instancia del repositorio de catálogo
    /// </summary>
    public CatalogRepository(IConfiguration configuration, ILogger<CatalogRepository> logger)
    {
        _logger = logger;
        _catalogFilePath = configuration["Catalog:FilePath"] ?? "./data/catalog/cars_extract.json";
        _csvFilePath = configuration["Catalog:CsvFilePath"] ?? "./src/Baba.Chatbot.Integrations/Catalog/sample_caso_ai_engineer.csv";
    }

    /// <summary>
    /// Busca vehículos en el catálogo aplicando los filtros especificados en el query
    /// Si no hay filtros, devuelve los primeros 10 vehículos
    /// </summary>
    public async Task<List<Vehicle>> SearchVehiclesAsync(VehicleQuery query, CancellationToken cancellationToken = default)
    {
        var allVehicles = await LoadVehiclesAsync(cancellationToken);

        if (!query.HasAnyFilter())
            return allVehicles.Take(10).ToList();

        var filtered = allVehicles.Where(v => MatchesQuery(v, query)).ToList();

        return filtered;
    }

    /// <summary>
    /// Obtiene un vehículo específico por su ID o stock_id
    /// </summary>
    public async Task<Vehicle?> GetVehicleByIdAsync(string id, CancellationToken cancellationToken = default)
    {
        var vehicles = await LoadVehiclesAsync(cancellationToken);
        return vehicles.FirstOrDefault(v => v.Id == id || v.StockId == id);
    }

    /// <summary>
    /// Carga los vehículos desde archivos CSV y JSON con caché en memoria
    /// </summary>
    private async Task<List<Vehicle>> LoadVehiclesAsync(CancellationToken cancellationToken)
    {
        if (_cachedVehicles != null)
            return _cachedVehicles;

        try
        {
            var vehicles = new List<Vehicle>();

            if (File.Exists(_csvFilePath))
            {
                _logger.LogInformation("Loading vehicles from CSV: {Path}", _csvFilePath);
                vehicles.AddRange(await LoadFromCsvAsync(_csvFilePath, cancellationToken));
            }

            if (File.Exists(_catalogFilePath))
            {
                _logger.LogInformation("Loading vehicles from JSON: {Path}", _catalogFilePath);
                vehicles.AddRange(await LoadFromJsonAsync(_catalogFilePath, cancellationToken));
            }

            if (vehicles.Count == 0)
            {
                _logger.LogWarning("No catalog files found or loaded");
            }

            _cachedVehicles = vehicles;
            _logger.LogInformation("Loaded {Count} vehicles from catalog", _cachedVehicles.Count);

            return _cachedVehicles;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error loading catalog");
            return new List<Vehicle>();
        }
    }

    /// <summary>
    /// Carga vehículos desde un archivo CSV
    /// Formato esperado: stock_id,km,price,make,model,year,version,bluetooth,largo,ancho,altura,car_play
    /// </summary>
    private async Task<List<Vehicle>> LoadFromCsvAsync(string filePath, CancellationToken cancellationToken)
    {
        var vehicles = new List<Vehicle>();
        
        using var reader = new StreamReader(filePath);
        
        var header = await reader.ReadLineAsync();
        if (header == null)
            return vehicles;

        int lineNumber = 1;
        while (!reader.EndOfStream)
        {
            var line = await reader.ReadLineAsync();
            if (string.IsNullOrWhiteSpace(line))
                continue;

            lineNumber++;
            
            try
            {
                var values = line.Split(',');
                if (values.Length < 8)
                {
                    _logger.LogWarning("Skipping line {Line}: insufficient columns", lineNumber);
                    continue;
                }

                var stockId = values[0].Trim();
                var km = ParseInt(values[1]);
                var price = ParseDecimal(values[2]);
                var make = values[3].Trim();
                var model = values[4].Trim();
                var year = ParseInt(values[5]) ?? DateTime.Now.Year;
                var version = values.Length > 6 ? values[6].Trim() : "";
                var bluetooth = values.Length > 7 ? ParseBool(values[7]) : null;
                var length = values.Length > 8 ? ParseInt(values[8]) : null;
                var width = values.Length > 9 ? ParseInt(values[9]) : null;
                var height = values.Length > 10 ? ParseInt(values[10]) : null;
                var carPlay = values.Length > 11 ? ParseBool(values[11]) : null;

                var vehicle = Vehicle.Create(
                    stockId,
                    make,
                    model,
                    year,
                    price ?? 0
                );

                vehicle.UpdateDetails(
                    version: version,
                    mileage: km,
                    length: length,
                    width: width,
                    height: height,
                    hasBluetooth: bluetooth,
                    hasCarPlay: carPlay,
                    stockId: stockId
                );

                vehicles.Add(vehicle);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Error parsing line {Line}", lineNumber);
            }
        }

        return vehicles;
    }

    /// <summary>
    /// Carga vehículos desde un archivo JSON
    /// </summary>
    private async Task<List<Vehicle>> LoadFromJsonAsync(string filePath, CancellationToken cancellationToken)
    {
        var json = await File.ReadAllTextAsync(filePath, cancellationToken);
        var vehicleData = JsonSerializer.Deserialize<List<VehicleDto>>(json);

        return vehicleData?.Select(dto => MapToEntity(dto)).ToList() ?? new List<Vehicle>();
    }

    /// <summary>
    /// Parsea un valor string a entero manejando valores nulos o "NA"
    /// </summary>
    private int? ParseInt(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Equals("NA", StringComparison.OrdinalIgnoreCase))
            return null;

        if (int.TryParse(value.Trim(), NumberStyles.Number, CultureInfo.InvariantCulture, out var result))
            return result;

        return null;
    }

    /// <summary>
    /// Parsea un valor string a decimal manejando valores nulos o "NA"
    /// </summary>
    private decimal? ParseDecimal(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Equals("NA", StringComparison.OrdinalIgnoreCase))
            return null;

        if (decimal.TryParse(value.Trim(), NumberStyles.Number, CultureInfo.InvariantCulture, out var result))
            return result;

        return null;
    }

    /// <summary>
    /// Parsea un valor string a booleano manejando valores nulos, "NA" o enteros (1/0)
    /// </summary>
    private bool? ParseBool(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Equals("NA", StringComparison.OrdinalIgnoreCase))
            return null;

        if (bool.TryParse(value.Trim(), out var result))
            return result;

        if (int.TryParse(value.Trim(), out var intValue))
            return intValue != 0;

        return null;
    }

    /// <summary>
    /// Verifica si un vehículo cumple con los criterios del query especificado
    /// </summary>
    private bool MatchesQuery(Vehicle vehicle, VehicleQuery query)
    {
        if (query.Brand != null && !vehicle.Brand.Contains(query.Brand, StringComparison.OrdinalIgnoreCase))
            return false;

        if (query.Model != null && !vehicle.Model.Contains(query.Model, StringComparison.OrdinalIgnoreCase))
            return false;

        if (query.MaxPrice.HasValue && vehicle.Price > query.MaxPrice.Value)
            return false;

        if (query.MinPrice.HasValue && vehicle.Price < query.MinPrice.Value)
            return false;

        if (query.MinYear.HasValue && vehicle.Year < query.MinYear.Value)
            return false;

        if (query.MaxYear.HasValue && vehicle.Year > query.MaxYear.Value)
            return false;

        if (query.MaxMileage.HasValue && vehicle.Mileage > query.MaxMileage.Value)
            return false;

        return true;
    }

    /// <summary>
    /// Mapea un DTO de JSON a la entidad de dominio Vehicle
    /// </summary>
    private Vehicle MapToEntity(VehicleDto dto)
    {
        var vehicle = Vehicle.Create(dto.id, dto.brand, dto.model, dto.year, dto.price);
        return vehicle;
    }

    /// <summary>
    /// DTO para deserializar vehículos desde JSON
    /// </summary>
    private record VehicleDto
    {
        public string id { get; init; } = string.Empty;
        public string brand { get; init; } = string.Empty;
        public string model { get; init; } = string.Empty;
        public int year { get; init; }
        public decimal price { get; init; }
        public int mileage { get; init; }
        public string transmission { get; init; } = string.Empty;
        public string fuelType { get; init; } = string.Empty;
        public string color { get; init; } = string.Empty;
        public string status { get; init; } = string.Empty;
    }
}

